src/mlpack/methods/cf/cf_impl.hpp

```diff
@@ -148,22 +132,271 @@ void CF::Train(const arma::sp_mat& data,
+  // neighbor search only on the H matrix, using the observation that if the
+  // rating matrix X = W*H, then d(X.col(i), X.col(j)) = d(W H.col(i), W
+  // H.col(j)).  This can be seen as nearest neighbor search on the H matrix
+  // with the Mahalanobis distance where M^{-1} = W^T W.  So, we'll decompose
+  // M^{-1} = L L^T (the Cholesky decomposition), and then multiply H by L^T.
+  // Then we can perform nearest neighbor search.
+  arma::mat l = arma::chol(w.t() * w);
+  arma::mat stretchedH = l * h; // Due to the Armadillo API, l is L^T.
+
+  // Now, we will use the decomposed w and h matrices to estimate what the user
+  // would have rated items as, and then pick the best items.
+
+  // Temporarily store feature vector of queried users.
+  arma::mat query(stretchedH.n_rows, users.n_elem);
+
+  // Select feature vectors of queried users.
+  for (size_t i = 0; i < users.n_elem; i++)
+    query.col(i) = stretchedH.col(users(i));
+
+  // Temporary storage for neighborhood of the queried users.
+  arma::Mat<size_t> neighborhood;
+
+  // Calculate the neighborhood of the queried users.  Note that the query user
+  // is part of the neighborhood---this is intentional.  We want to use an
+  // average of both the query user and the local neighborhood of the query
+  // user.
+  // The neighbor search technique should be a template parameter.
+  neighbor::KNN a(stretchedH);
+  arma::mat resultingDistances; // Temporary storage.
+  a.Search(query, numUsersForSimilarity, neighborhood, resultingDistances);
+
+  // Generate recommendations for each query user by finding the maximum numRecs
+  // elements in the averages matrix.
+  recommendations.set_size(numRecs, users.n_elem);
+  arma::mat values(numRecs, users.n_elem);
+  recommendations.fill(SIZE_MAX);
+  values.fill(DBL_MAX);
+
+  for (size_t i = 0; i < users.n_elem; i++)
+  {
+    // First, calculate average of neighborhood values.
+    arma::vec averages;
+    averages.zeros(cleanedData.n_rows);
+
+    for (size_t j = 0; j < neighborhood.n_rows; ++j)
+      averages += w * h.col(neighborhood(j, i));
+    averages /= neighborhood.n_rows;
+
+    // Let's build the list of candidate recomendations for the given user.
+    // Default candidate: the smallest possible value and invalid item number.
+    const Candidate def = std::make_pair(-DBL_MAX, cleanedData.n_rows);
+    std::vector<Candidate> vect(numRecs, def);
+    typedef std::priority_queue<Candidate, std::vector<Candidate>, CandidateCmp>
+        CandidateList;
+    CandidateList pqueue(CandidateCmp(), std::move(vect));
+
+    // Look through the averages column corresponding to the current user.
+    for (size_t j = 0; j < averages.n_rows; ++j)
+    {
+      // Ensure that the user hasn't already rated the item.
+      if (cleanedData(j, users(i)) != 0.0)
+        continue; // The user already rated the item.
+
+      // Is the estimated value better than the worst candidate?
+      if (averages[j] > pqueue.top().first)
```

Comment:
I think you should denormalize `avereges[j]` first since `pqueue.top().first` is a denormalized value.